### PR TITLE
Improved modem functionality

### PR DESCRIPTION
**Pull request by ArendJanKramer**
This PR features:
 - The ability to setup communication with car, according to ISO15118 standard
 - Work together with pyPLC (https://github.com/uhi22/pyPLC) which can handle the digital communication with the car, and submit the SoC states back to SmartEVSE
 - Compute the current SoC, based on the EV meter
 - Reliably reset SoC values upon disconnect with vehicle
 - "Re-plug" the vehicle if the communication fails, and after communication was succesfull to resume AC charging
 - Display the SoC values on the web UI

Work todo:
 - Set the battery capacity of the car via the menu, or via API. Currently hardcoded on 28kWh for Ioniq vehicle
 - Feature to stop charging once a specified SoC has been reached
 - Authorization of car based on MAC address of car

Please review and merge pr, rather than copy/pasting code. IMO git is not meant for copy-paste.
I'm happy to make changes or to add code/docs where needed.

### CODE REVIEW

File: SmartEVSE-3/data/index.html
Comment (ArendJanKramer) [inline SmartEVSE-3/data/index.html:1]: @dingo35 What formatter do you use? VS Code reformatted the file, which is a pain to undo by hand.

Comment (dingo35) [reply SmartEVSE-3/data/index.html:1]: I use no formatter to prevent those massive, unnecessary changes that make git-blame useless...

Comment (ArendJanKramer) [reply SmartEVSE-3/data/index.html:1]: @dingo35 How do you want to address this?
I have two different options:
 - I can undo the formatting I accidentally did, bit of work but not impossible. Have to do some git resets in order to get history back.
 - You can format the file first in a merge with a formatter of your choice, and my changes can be merged after. Leaves the formatter changes to your name. I can rebase if conflicts are made, and use the same formatter.

Usually it's a good practice to agree on common formatters/linters and enforce them in the ci using style checks (e.g. github actions) and agree on those; but I also understand the project is incredible small with only a handful of contributors.
In the end it's your repo, so I leave it up to you, I'll adjust.

Comment (dingo35) [reply SmartEVSE-3/data/index.html:1]: Im a bit busy coming days, I hope to be able to take a look at all the outstanding PR's and bug reports at the end of the week, Ill get back to you! Appreciate your contribution and cooperative attitude!

Comment (dingo35) [reply SmartEVSE-3/data/index.html:1]: First of all: great work!! Code looks great, some small stuff but two more import things:
1) why do we need the 28kW hardcoded? Of course it would be easy to make a menu item out of this, but it wouldn't make any sense since any other car could connect to the SmartEVSE with a different battery pack capacity. The goal was to get information out of the car with the modem, _not_ to have to insert more info to the SmartEVSE :-) 
But seriously, could we get rid of this, or is this "work in progress" to discover the protocol, so it can deliver the total battery capacity as it can with the SoC?
2) You have ModemTimer1 thru 3 defined; would it be possible to either document them better, or, preferable, give them more "logical" names to distinguish between them?

I have some nitty-bitty stuff in the code which I will try to edit/comment inline if I have more time.

Comment (ArendJanKramer) [reply SmartEVSE-3/data/index.html:1]: @dingo35 I have un-done the indentations of index.html. 

1. I also factored out the SoC code.
For now hard-coded, but I can open an issue and assign it to myself for using the modem. We need to adjust pyPLC as well to extract and submit the information, so it requires a bit of effort. But removing it for now would mean double work, so wasn't keen on that. Not sure how to deal with cars that do not submit the info though. Most people charge their own car(s), so it'd make sense for them at least to have fallback. But as long the small pool of modem owners are having the info, perhaps we can just ignore that use-case.

Another interesting consideration is battery health. Ioniq reports 28kWh on the modem, but since my car has low mileage, my charge goes beyond 100% in theory (105-110 or so, depending on start point) when I work with 28kWh in the equation. So in practice the battery is slightly bigger, and I should compensate for this in the formula, but this depends on state of health and other variables per car, which we don't have access to. A universal workaround would be pausing the charge at exactly 10kWh, readout the car again and compute delta in charge, but that solution seems a bit excessive to me tbh. 

2. Done, improved names and added clarification.

File: SmartEVSE-3/data/index.html
Comment (dingo35) [inline SmartEVSE-3/data/index.html:1]: Reformatting change, please unformat

File: SmartEVSE-3/data/index.html
Comment (dingo35) [inline SmartEVSE-3/data/index.html:38]: Reformatting change, please unformat

File: SmartEVSE-3/data/index.html
Comment (dingo35) [inline SmartEVSE-3/data/index.html:60]: Reformatting change, please unformat

File: SmartEVSE-3/data/index.html
Comment (dingo35) [inline SmartEVSE-3/data/index.html:359]: Reformatting change, please unformat

File: SmartEVSE-3/data/index.html
Comment (dingo35) [inline SmartEVSE-3/data/index.html:519]: Reformatting change, please unformat

File: SmartEVSE-3/data/index.html
Comment (dingo35) [inline SmartEVSE-3/data/index.html:825]: Reformatting change, please unformat

File: SmartEVSE-3/data/index.html
Comment (dingo35) [inline SmartEVSE-3/data/index.html:788]: Reformatting change, please unformat

File: SmartEVSE-3/src/evse.cpp
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:1686]: Reformatting change, please unformat

File: SmartEVSE-3/src/evse.cpp
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:1691]: Reformatting change, please unformat

File: SmartEVSE-3/src/evse.cpp
Comment (dingo35) [inline SmartEVSE-3/src/evse.cpp:1768]: Reformatting change, please unformat

**Review by dingo35 — CHANGES_REQUESTED** [review]
Just some minor formatting changes requested; I would have done it myself but I'm not sure if this would interfere with your personal fork, as you said we must use github to the optimum!

Thanks for your great work Arend Jan!

File: SmartEVSE-3/data/index.html
Comment (ArendJanKramer) [reply SmartEVSE-3/data/index.html:1]: @dingo35 I undid most formatting, but both VS Code and PyCharm do not accept blank spaces after html tags, so I can't reasonably undo those, unless I resort back to a text editor. C++ code was doable though.
For consistent white space between lines, I'm fine. But as I'm not planning to do further developments in a text editor, they likely will come back. 

I'd suggest to pick a formatter + config, I can't care less which one, but it'd lower the threshold to make changes.

Comment (dingo35) [reply SmartEVSE-3/data/index.html:1]: @ArendJanKramer  I must overlook something, but none of your formatting changes made it into the "Changed Files" part of this PR. This is taking up too much of both of our times, so I took your PR, git add -p divided it into nice chunks and 10 minutes later it was all in there. I know you would have preferred github, but this tool becomes more of a handicap then a help...

Resyncing it with your personal repo should only give a few formatting changes, so shouldn't take up much more of your time either.
As to data/index.html, I agree that is a nightmare right now. I'm looking for a good WYSIWIG kind of editor that solves al those troubles, but all I can find / is recommended to me are "framework A" or "framework B"; I believe this index.html stems from the "bootstrap" framework, changing framework doesn't seem like a solution to me, and using formatters "cleans up" the code until the next proposition for the next IDE/formatter/tool comes up. And it all breaks up the code history, which is very important to me since there are parts of the code that I do not fully understand (yet) and I want to keep as much compatiblitiy with future Stegen developments as possible...

Hope you understand, look forward to working further with you! Hans.

Comment (ArendJanKramer) [reply SmartEVSE-3/data/index.html:1]: @dingo35 It's not about Github, it's about git history and future developments.
I mean, you didn't like the formatting changes because it would change the git history, but now 100% of my changes are gone in the git blame, so what has been resolved exactly?

There are two things related to formatting:
 -- Reindenting and restructuring files: Sure, I agree this is not supposed to happen, I undid these
 -- Left behind spaces, such as after html tags and white spaces that are more than usually

I can't reasonably contribute anymore to the html parts, unless I resort back to text editors. To me that's strange, platformio uses VS Code, and the Jetbrains producs aren't exactly unheard of either. 

We don't have to resort or pin to any of the editors, it can be as simple as:
 -- No more than two consecutive blank lines in html
 -- Reindent is two spaces per tag
 -- A line should never end with a space

You can see the formatting rules of google for instance aren't demanding:
https://developers.google.com/style/html-formatting
Please note the "Don't leave trailing spaces", is what most editors will undo.

Comment (dingo35) [reply SmartEVSE-3/data/index.html:1]: Happy to look for a solution where your changes are preserved, but with a standard "squash and merge" they arent preserved either, are they? And when I dont squash them, but rebase them, the history is still disturbed by formatting/unformatting commits...

As to the html, it is so little code, it doesnt matter how you deliver your changes, Im happy to incorporate them by hand. The essence is in the .cpp stuff IMHO....

Comment (ArendJanKramer) [reply SmartEVSE-3/data/index.html:1]: @dingo35 For the changes purely, I re-commited your last commit in my fork.
So by syncing to that branch that part should be resolved.

For the html I'm having some ideas to develop a few more things, such as vehicle profiles.
My idea is you can add some cars, with respective serial numbers for authorization (mac addresses, like fastned does: https://presspage-production-content.s3.amazonaws.com/uploads/2519/autocharge-whitepaper-customerversion.pdf?10000) but also being able to set battery capacities if the car does not report it for charge tracking.
Those things are still in the design phase though.

Other things I already have integrated, but not shared yet is the ability to stop charging at configured SoC point, to keep battery more healthy. 

For me trailing spaces are the worst, so if you can remove those in some PR I'd be extremely happy. At least it'd bring some consistency, as only a few lines seem to have them.
(The stegen branch doesn't seem to have them as much, so it wouldn't break things with him fortunately)

I'm even happy to create a Github actions for you to do a an automated style check on each PR.

**Comment by dingo35** [comment]
@ArendJanKramer I'm afraid I'm going to have to revert your code; it prevents charging when Modem = Not present, so it looks like you never even tested without modem/pyplc present?

**Comment by ArendJanKramer** [comment]
> @ArendJanKramer I'm afraid I'm going to have to revert your code; it prevents charging when Modem = Not present, so it looks like you never even tested without modem/pyplc present?

It works fine here? 
Normal, smart and solar seem to work, although solar is waiting as its dark here.

**Comment by dingo35** [comment]
Perhaps EV dependent? My EV didnt charge last evening, didnt have time to look at it, reverted theast commit and there it went.
No EV here today so I cant try....

**Comment by dingo35** [comment]
@ArendJanKramer False alarm, I cannot reproduce the problem, it charges fine now. Might have been the EV, there are quite some bugs in my cars firmware and VAG has problems getting their software development trajectory right.... Still waiting for the firmware update that will enable modem communication....

**Review by dingo35 — APPROVED** [review]

**Pull request merged**
